File: src/Tests/Behavioral/StructPromotion/StructPromotion.cs
namespace go;

using fmt = fmt_package;
using strings = strings_package;

partial class main_package {

[GoType] partial struct Person {
    internal @string name;
    internal int32 age;
}

public static bool IsDr(this Person p) {
    return strings.HasPrefix(p.name, "Dr"u8);
}

public static bool IsAdult(this Person p) {
    return p.age >= 18;
}

[GoType] partial struct Employee {
    internal @string position;
}

public static bool IsManager(this Employee e) {
    return e.position == "manager"u8;
}

[GoType] partial struct Record {
    public partial ref Person Person { get; }
    public partial ref Employee Employee { get; }
}

public static bool IsDr(this Record p) {
    return strings.HasPrefix(p.name, "Dr"u8) && p.age > 18;
}

[GoType] partial struct commonBase {
    internal @string tag;
}

[GoRecv] internal static @string describe(this ref commonBase c) {
    return "base:"u8 + c.tag;
}

[GoType] partial struct ledger {
    internal partial ref commonBase commonBase { get; }
    internal nint seq;
}

internal static @string describe(this ж<ledger> Ꮡl) {
    ref var l = ref Ꮡl.Value;

    var p = Ꮡl.of(ledger.Ꮡseq);
    p.Value++;
    return "ledger:"u8 + l.tag;
}

internal static void Main() {
    var person = new Person(name: "Dr. Michał"u8, age: 29);
    fmt.Println(person);
    fmt.Println(person.IsDr());
    var record = new Record(nil);
    record.name = "Dr. Michał"u8;
    record.age = 18;
    record.position = "software engineer"u8;
    fmt.Println(record);
    fmt.Println(record.name);
    fmt.Println(record.age);
    fmt.Println(record.position);
    fmt.Println(record.IsAdult());
    fmt.Println(record.IsManager());
    fmt.Println(record.IsDr());
    var l = Ꮡ(new ledger(nil));
    l.Value.tag = "x"u8;
    fmt.Println(l.describe(), l.describe(), (~l).seq);
}

} // end main_package

File: src/Tests/Behavioral/StructPromotion/StructPromotion.go
package main

import (
	"fmt"
	"strings"
)

type Person struct {
	name string
	age  int32
}

func (p Person) IsDr() bool {
	return strings.HasPrefix(p.name, "Dr")
}

func (p Person) IsAdult() bool {
	return p.age >= 18
}

type Employee struct {
	position string
}

func (e Employee) IsManager() bool {
	return e.position == "manager"
}

type Record struct {
	Person
	Employee
}

func (p Record) IsDr() bool {
	return strings.HasPrefix(p.name, "Dr") && p.age > 18
}

// commonBase carries a POINTER-receiver method `describe`. Since it does not take the address
// of a receiver field it is emitted as a value-ref extension, so promotion into an embedder
// generates BOTH a value and a box overload.
type commonBase struct {
	tag string
}

func (c *commonBase) describe() string {
	return "base:" + c.tag
}

// ledger embeds commonBase and declares its OWN pointer-receiver `describe` that takes the
// address of a receiver field (`&l.seq`) — the converter emits it ONLY as a box primary
// (`describe(this ж<ledger> …)`). That box form is invisible to the value-receiver-only
// promotion-shadowing check, so commonBase.describe's promoted BOX overload used to duplicate
// it (CS0111 — encoding/gob's structType embedding CommonType, both declaring string()).
type ledger struct {
	commonBase
	seq int
}

func (l *ledger) describe() string {
	p := &l.seq
	*p++
	return "ledger:" + l.tag // reads the promoted commonBase.tag field
}

func main() {
	person := Person{name: "Dr. Michał", age: 29}
	fmt.Println(person)        // {Dr. Michał 29}
	fmt.Println(person.IsDr()) // true

	record := Record{}
	record.name = "Dr. Michał"
	record.age = 18
	record.position = "software engineer"

	fmt.Println(record)             // {{Dr. Michał 18} {software engineer}}
	fmt.Println(record.name)        // Dr. Michał
	fmt.Println(record.age)         // 18
	fmt.Println(record.position)    // software engineer
	fmt.Println(record.IsAdult())   // true
	fmt.Println(record.IsManager()) // false
	fmt.Println(record.IsDr())      // false

	l := &ledger{}
	l.tag = "x"
	fmt.Println(l.describe(), l.describe(), l.seq) // ledger:x ledger:x 2
}
